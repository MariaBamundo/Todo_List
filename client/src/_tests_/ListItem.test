import React from 'react';
import {ApolloProvider} from 'react-apollo';
import ReactDOM from 'react-dom';
import ListItem from '../components/ListItem';
import renderer from 'react-test-renderer';
import {ApolloClient} from 'apollo-client';
import {HttpLink} from 'apollo-link-http';
import fetch from 'node-fetch';
import {InMemoryCache} from 'apollo-cache-inmemory';

/** Test structure**/
//Setup initial State
//Dispatch Action
//Expect Data to Change

const client = new ApolloClient({
    link: new HttpLink({
        uri: 'http://localhost:4000/graphql',
        fetch,
    }),
    cache: new InMemoryCache()
})

//A test to see if the app renders
describe('ListItem', () => {
    it('should render without crashing', () => {
        const rendered = renderer.create(
            <ApolloProvider client={client}>
                <ListItem />
            </ApolloProvider>
        );
        expect(rendered).toBeTruthy();
    });
});

//Test render
//Test update