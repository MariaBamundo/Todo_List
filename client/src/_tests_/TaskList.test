import React from 'react';
import renderer from 'react-test-renderer';

import TaskList from '../components/TaskList';

//Test render


//Test checkbox
//Test delete
//Test add