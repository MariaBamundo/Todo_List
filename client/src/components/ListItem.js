import React, { Component } from 'react';
import { graphql} from 'react-apollo';
import { UPDATE_TASK } from '../mutations';

//Class for displaying a single list item
class ListItem extends Component {
    //constructor sets states using props and event handlers
    constructor(props) {
        super(props);
        this.state = {
            id: this.props.task.id,
            name: this.props.task.name,
            editing: false
        };
        this.handleChange = this.handleChange.bind(this);
    }

    //This method sets the state of editing to true 
    //so that a text box is displayed for updating an individual task
    editTask() {
        this.setState({
            editing: true
        });
    }

    //This method is called when the user hits save after editing a task.
    saveTask() {
        //call th updateTask mutator method to make the change in the database
        this.props.updateTask({
            variables: {
                id: this.state.id,
                name: this.state.name
            },
            //set the state of the tasks name to the new name so it renders without a page reload
            update: (store, { data: { updateTask } }) => {
                this.setState({
                    name: this.state.name,
                    editing: false
                }); 
            }

        });

    }

    //Called when a user types in the input displayed when editing==true
    handleChange(e) {
        //Set the state of name to whatever the user is typeing so it is visible in the input
        this.setState({ name: e.target.value });
    }


    //Called when the page renders
    showTask() {
        //If a user is editing a list item 
        if (this.state.editing === true) {
            return (
                <span>
                    <input
                        value={this.state.name} size={4} onChange={this.handleChange} />
                    <button className="save" onClick={() => this.saveTask()}>Save</button>
                </span>
            );
        }
        else {
            return (
                <span id={this.state.id}>
                    {this.state.name}
                    <button className="edit" onClick={() => this.editTask()}>Edit</button>
                </span>
            );
        }
    }

    //call show task
    render() {
        return (this.showTask());
    }
}

//export the update method and the entire class
export default graphql(UPDATE_TASK, { name: 'updateTask' })(ListItem);