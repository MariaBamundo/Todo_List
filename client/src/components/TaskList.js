import React from 'react';
import { graphql, compose } from 'react-apollo';

//import mutations & queries
import { ADD_TASK } from '../mutations';
import { COMPLETE_TASK } from '../mutations';
import { DELETE_TASK } from '../mutations';
import { GET_TASKS } from '../queries';

import ListItem from './ListItem';

class TaskList extends React.Component {

    //Constuctor sets state and event handlers
    constructor(props) {
        super(props);
        this.state = {
            items: this.props.items,
            editing: false,
            name: ''
        };
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }
    //change state.name variable. Called when the user types in the input
    handleChange(e) {
        this.setState({ name: e.target.value });
    }

    //Called when the user submits the form
    handleSubmit(e) {
        e.preventDefault();
        //If a user attempts to submit nothing just return 
        if (!this.state.name.length) {
            return;
        }
        //use mutator method to update the database
        this.props.addNewTask({
            variables: {
                name: this.state.name,
            },
            //set the state to rerender the TaskList with the new task included
            update: (store, { data: { addTask } }) => {
                this.setState({
                    name: '',
                    items: [...this.state.items, addTask]
                });
            },
        }).then(function handleChange(response) {
            console.log(response);
        });
    }

    //Called when the user hits the delete button
    deleteTask(item) {
        //use mutator method to update the database
        this.props.removeTask({
            variables: {
                id: item.id,
            },
        }).then(function getResponse(response) {
            console.log(response);
        });
        //set the state to the list with non deleted items
        this.setState(prevState => ({
            items: prevState.items.filter(task => task != item)
        }));
    }


    checkTask(item) {
        //call mutator method to update database
        this.props.completeTask({
            variables: {
                id: item.id,
                //set isdone the NOT isDone
                isDone: !item.isDone,
            },
            update: (store, { data: { completeTask } }) => {
                //get the array of items
                const data = this.state.items.slice();
                //get index of completed task
                const index = data.findIndex(item => item.id === completeTask.id);
                //Update the isDone of completed task on the front end
                data[index] = completeTask;
                //set state so the isDone field corresponds with database without reloading page
                this.setState({
                    items: data
                })
            }
        }).then(function getResponse(response) {
            console.log(response);
        })
    }

    //Render the list of Tasks
    render() {

        return (

            //Input bar at the top of the page uses event handlers to addTask
            <div className='taskList'>
                <form onSubmit={this.handleSubmit}>
                    <div className='input'>
                        <input
                            className="addTask"
                            onChange={this.handleChange}
                            value={this.state.name}
                            placeholder={'Add a task'}
                        />
                    </div>
                </form>

                <ul className="tasks">
                    {
                        //Map the List of individual items 
                        //Use the ListItem class to display the item name and its edit functionality
                        //Add a checkbox and delete button to use delete and check mutator methods
                        this.state.items.map(item => {
                            return <li key={item.id}>
                                <input type="checkbox" checked={item.isDone} onChange={() => this.checkTask(item)}></input>
                                <ListItem task={item} />
                                <button onClick={() => this.deleteTask(item)}>
                                    {'Delete'}
                                </button>
                            </li>

                        })

                    }
                </ul>
            </div>

        );
    }
}

//Export all querys and mutations used in this class aswell as the class itself
export default compose(graphql(GET_TASKS, { name: 'getTasks' }),
    graphql(DELETE_TASK, { name: 'removeTask' }),
    graphql(COMPLETE_TASK, { name: 'completeTask' }),
    graphql(ADD_TASK, { name: 'addNewTask' }))
    (TaskList);