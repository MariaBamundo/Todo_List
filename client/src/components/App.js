import React, { Component } from 'react';
import { graphql, compose } from 'react-apollo';

import './App.css';

import { GET_TASKS } from '../queries';

import TaskList from "./TaskList";

//create app
class App extends Component {
    //just props no state declaration required
    constructor(props) {
        super(props)
    }

    //render the app using TaskList
    render() {
        const loading = this.props.getTasks.loading;
        if (loading) {
            return <div> loading </div>;
        }
        return (
            <div className="App">
                <h1>Todo...</h1>
                <TaskList items={this.props.getTasks.tasks} />
            </div>
        );
    }
}

export default compose(graphql(GET_TASKS, { name: 'getTasks' }))
    (App);


