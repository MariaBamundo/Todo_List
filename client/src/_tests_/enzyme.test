import React from 'react';
import App from '../components/App';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { shallow, mount, render } from 'enzyme';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('App Component', () => {
    it('should render without throwing an error', () => {
        expect(shallow(<App />).exists(<div className="App"></div>)).toBe(true)
        /*App renders all elements*//*
        it('renders all parts', () => {
            const props = [
                {
                    "id": "1",
                    "name": "temp",
                    "isDone": false
                }
            ];
    
            const wrapper = shallow(<App {...props} />);
    
            expect(wrapper.find(<div className="App" />).exists()).toBe(true);
            
            expect(wrapper.find(GreetWrapper).exists()).toBe(true);
            expect(wrapper.find('span').text()).toBe('welcome John Doe');*/
    })
    it('renders correctly', () => {
        const rendered = renderer.create(
            <ApolloProvider client={client}>
                <App />
            </ApolloProvider>
        );
        expect(rendered).toMatchSnapshot();
    });
})