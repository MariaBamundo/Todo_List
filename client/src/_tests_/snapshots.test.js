import React from 'react';
import { ApolloProvider } from 'react-apollo';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { ApolloClient } from 'apollo-client';
import { HttpLink } from 'apollo-link-http';
import fetch from 'node-fetch';
import { InMemoryCache } from 'apollo-cache-inmemory';

//import components to be tested
import App from '../components/App';
import TaskList from '../components/TaskList';
import ListItem from '../components/ListItem';

//enzyme imports
import { shallow, mount, render } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
/** Test structure**/
//Setup initial State
//Dispatch Action
//Expect Data to Change

const client = new ApolloClient({
    link: new HttpLink({
        uri: 'http://localhost:4000/graphql',
        fetch,
    }),
    cache: new InMemoryCache()
})

//Test Data
const mockedItem = { id: 'item', name: 'item', isDone: false };

const mockedTasks = [
    { id: 'task1', name: 'task1', isDone: false },
    { id: 'task2', name: 'task2', isDone: true }
];


describe('Use Mock Data', () => {
    test('App', () => {
        const component = renderer.create(
            <ApolloProvider client={client}>
                <App />
            </ApolloProvider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        //console.log(tree);
        //expect(component.contains('div')).toBe(true);
    });


    test('TaskList', () => {
        const component = renderer.create(
            <ApolloProvider client={client}>
                <TaskList items={mockedTasks} />
            </ApolloProvider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        //console.log(tree);

        // manually trigger functions

        console.log(tree.children[1].children[0].children[0].props.onChange());
        console.log(component);
        // re-rendering
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        

        //test input use classname taskList

        //test delete

    });
    /*
        test('ListItem', () => {
            const component = renderer.create(
                <ApolloProvider client={client}>
                    <ListItem task={mockedItem} />
                </ApolloProvider>
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
            console.log(tree);
        });*/
});
