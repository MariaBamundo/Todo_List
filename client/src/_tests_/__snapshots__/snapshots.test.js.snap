// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Use Mock Data App 1`] = `
<div>
   loading 
</div>
`;

exports[`Use Mock Data TaskList 1`] = `
<div
  className="taskList"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="input"
    >
      <input
        className="addTask"
        onChange={[Function]}
        placeholder="Add a task"
        value=""
      />
    </div>
  </form>
  <ul
    className="tasks"
  >
    <li>
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        id="task1"
      >
        task1
        <button
          className="edit"
          onClick={[Function]}
        >
          Edit
        </button>
      </span>
      <button
        onClick={[Function]}
      >
        Delete
      </button>
    </li>
    <li>
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        id="task2"
      >
        task2
        <button
          className="edit"
          onClick={[Function]}
        >
          Edit
        </button>
      </span>
      <button
        onClick={[Function]}
      >
        Delete
      </button>
    </li>
  </ul>
</div>
`;

exports[`Use Mock Data TaskList 2`] = `
<div
  className="taskList"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="input"
    >
      <input
        className="addTask"
        onChange={[Function]}
        placeholder="Add a task"
        value=""
      />
    </div>
  </form>
  <ul
    className="tasks"
  >
    <li>
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        id="task1"
      >
        task1
        <button
          className="edit"
          onClick={[Function]}
        >
          Edit
        </button>
      </span>
      <button
        onClick={[Function]}
      >
        Delete
      </button>
    </li>
    <li>
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        id="task2"
      >
        task2
        <button
          className="edit"
          onClick={[Function]}
        >
          Edit
        </button>
      </span>
      <button
        onClick={[Function]}
      >
        Delete
      </button>
    </li>
  </ul>
</div>
`;
